File: java/json-transform/src/main/java/ai/bizone/jsontransform/functions/common/TransformerFunction.java
package ai.bizone.jsontransform.functions.common;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * Base class for all transformer functions.
 */
public abstract class TransformerFunction {
    private final Map<String, Object> defaultValues;
    private final FunctionDescription description;

    public TransformerFunction(FunctionDescription description) {
        this.description = description;
        this.defaultValues = new HashMap<>();
        var args = description.getArguments();
        for (var arg : args.keySet()) {
            this.defaultValues.put(arg, args.get(arg).defaultValue);
        }
    }

    public TransformerFunction() {
        this(FunctionDescription.of(Collections.emptyMap()));
    }

    /**
     * Apply the function to the given context.
     * @param context the context
     * @return the result of the function
     */
    public abstract Object apply(FunctionContext context);

    /**
     * Get the argument type for the given name.
     * @param name the name of the argument (null will return the primary argument)
     * @return the argument type or null if not found
     */
    public ArgumentType getArgument(String name) {
        if (name == null) return null;
        return description.getArguments().get(name);
    }

    /**
     * Get the arguments for this function.
     * @return the function arguments
     */
    public Map<String, ArgumentType> getArguments() {
        return description.getArguments();
    }

    /**
     * Get the default value for the given argument name.
     * @param name the argument name
     * @return the default value or null if not found
     */
    public Object getDefaultValue(String name) {
        if (name == null) return null;
        return defaultValues.get(name);
    }
}

File: javascript/json-transform/src/functions/common/TransformerFunction.ts
import FunctionContext from "./FunctionContext";
import { ArgumentsSet, FunctionDescription } from "./FunctionDescription";

function areSame(a: any, b: any): boolean {
  if (a === b) return true;
  if (typeof a !== typeof b) return false;
  if (typeof a === "string" && typeof b === "string") {
    return a.toLowerCase() === b.toLowerCase();
  }
  return false;
}
/**
 * Base class for all transformer functions.
 */
abstract class TransformerFunction {
  protected readonly description: FunctionDescription;

  protected constructor(description: FunctionDescription) {
    this.description = description;
  }

  public parseArguments(args?: Record<string, any> | (string | null)[]): ArgumentsSet | undefined {
    const argsSets = this.description.argsSets;
    if (!argsSets) return undefined;
    if (argsSets.length === 1) return argsSets[0];
    if (!args) return argsSets.at(-1);
    const argSet = Array.isArray(args)
      ? argsSets.find(x =>
          x.every(
            (arg, i) => (!arg.const || areSame(arg.const, args[i])) && (!arg.exists || typeof args[i] !== "undefined"),
          ),
        )
      : argsSets.find(x =>
          x.every(
            arg =>
              (!arg.const || areSame(arg.const, args[arg.name])) &&
              (!arg.exists || typeof args[arg.name] !== "undefined"),
          ),
        );
    return argSet ?? argsSets.at(-1);
  }

  /**
   * Apply the function to the given context.
   * @param context the context
   * @return the result of the function
   */
  public abstract apply(context: FunctionContext): Promise<any>;

  /**
   * Check if the function allows arguments to be used as input.
   * @return true if arguments can be used as input, false otherwise
   */
  public allowsArgumentsAsInput() {
    return this.description.allowsArgumentsAsInput;
  }

  /**
   * Check if the function allows arguments to be used as input.
   * @return true if arguments can be used as input, false otherwise
   */
  public inputIsRaw() {
    return this.description.inputIsRaw;
  }
}

export default TransformerFunction;
